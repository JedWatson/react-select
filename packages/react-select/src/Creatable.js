// @flow

import React, {
  Component,
  type Config,
  type Node,
  type AbstractComponent,
  type ElementRef,
  type ElementConfig,
} from 'react';
import Select, { type Props as SelectProps } from './Select';
import type { OptionType, OptionsType, ValueType, ActionMeta, InputActionMeta } from './types';
import { cleanValue } from './utils';

export type DefaultCreatableProps = {|
  /* Allow options to be created while the `isLoading` prop is true. Useful to
     prevent the "create new ..." option being displayed while async results are
     still being loaded. */
  allowCreateWhileLoading: boolean,
  /* Sets the position of the createOption element in your options list. Defaults to 'last' */
  createOptionPosition: 'first' | 'last',
  /* Gets the label for the "create new ..." option in the menu. Is given the
     current input value. */
  formatCreateLabel: string => Node,
  /* Determines whether the "create new ..." option should be displayed based on
     the current input value, select value and options array. */
  isValidNewOption: (string, OptionsType, OptionsType) => boolean,
  /* Returns the data for the new option when it is created. Used to display the
     value, and is passed to `onChange`. */
  getNewOptionData: (string, Node) => OptionType,
|};
export type CreatableProps = {
  ...DefaultCreatableProps,
  /* If provided, this will be called with the input value when a new option is
     created, and `onChange` will **not** be called. Use this when you need more
     control over what happens when new options are created. */
  onCreateOption?: string => void,
  /* Sets the position of the createOption element in your options list. Defaults to 'last' */
  createOptionPosition: 'first' | 'last',
  /* Name of the HTML Input (optional - without this, no input will be rendered) */
  name?: string,
  options?: OptionsType,
  inputValue: string,
  value: ValueType,
  isLoading?: boolean,
  isMulti?: boolean,
  onChange: (ValueType, ActionMeta) => void,
  onInputChange: (string, InputActionMeta) => void,
};

export type Props = SelectProps & CreatableProps;

const compareOption = (inputValue = '', option) => {
  const candidate = String(inputValue).toLowerCase();
  const optionValue = String(option.value).toLowerCase();
  const optionLabel = String(option.label).toLowerCase();
  return optionValue === candidate || optionLabel === candidate;
};

const builtins = {
  formatCreateLabel: (inputValue: string) => `Create "${inputValue}"`,
  isValidNewOption: (
    inputValue: string,
    selectValue: OptionsType,
    selectOptions: OptionsType
  ) =>
    !(
      !inputValue ||
      selectValue.some(option => compareOption(inputValue, option)) ||
      selectOptions.some(option => compareOption(inputValue, option))
    ),
  getNewOptionData: (inputValue: string, optionLabel: Node) => ({
    label: optionLabel,
    value: inputValue,
    __isNew__: true,
  }),
};

export const defaultProps: DefaultCreatableProps = {
  allowCreateWhileLoading: false,
  createOptionPosition: 'last',
  ...builtins,
};

type State = {
  inputValue: string,
};

export const makeCreatableSelect = <C: {}>(
  SelectComponent: AbstractComponent<C>
): AbstractComponent<C & Config<CreatableProps, DefaultCreatableProps>> =>
  class Creatable extends Component<CreatableProps & C, State> {
    static defaultProps = defaultProps;
    select: ElementRef<*>;
    buildOptions = createMemoizedOptionsBuilder();
    state = {
      inputValue: '',
    };
    static getDerivedStateFromProps(props: CreatableProps, state: State) {
      if (inputIsControlled(props) && props.inputValue !== state.inputValue) {
        return {
          inputValue: props.inputValue
        };
      }

      return null;
    }
    onChange = (newValue: ValueType, actionMeta: ActionMeta) => {
      const {
        getNewOptionData,
        inputValue,
        isMulti,
        onChange,
        onCreateOption,
        value,
        name,
      } = this.props;
      if (actionMeta.action !== 'select-option') {
        return onChange(newValue, actionMeta);
      }
      const newOption = this.buildOptions.getNewOption();
      const valueArray = Array.isArray(newValue) ? newValue : [newValue];

      if (valueArray[valueArray.length - 1] === newOption) {
        if (onCreateOption) onCreateOption(inputValue);
        else {
          const newOptionData = getNewOptionData(inputValue, inputValue);
          const newActionMeta = { action: 'create-option', name };
          if (isMulti) {
            onChange([...cleanValue(value), newOptionData], newActionMeta);
          } else {
            onChange(newOptionData, newActionMeta);
          }
        }
        return;
      }
      onChange(newValue, actionMeta);
    };
    onInputChange = (value: string, actionMeta: InputActionMeta) => {
      if (inputIsControlled(this.props)) {
        if (typeof this.props.onInputChange === 'function') {
          this.props.onInputChange(value, actionMeta);
        }
      } else {
        this.setState({
          inputValue: value,
        });
      }
    };
    focus() {
      this.select.focus();
    }
    blur() {
      this.select.blur();
    }
    render() {
      return (
        <SelectComponent
          {...this.props}
          ref={ref => {
            this.select = ref;
          }}
          options={this.buildOptions({ ...this.state, ...this.props })}
          onChange={this.onChange}
          onInputChange={this.onInputChange}
          inputValue={inputIsControlled(this.props) ? this.props.inputValue : this.state.inputValue}
        />
      );
    }
  };

// TODO: do this in package entrypoint
export default makeCreatableSelect<ElementConfig<typeof Select>>(
  Select
);

const inputIsControlled = (props: CreatableProps): boolean => typeof props.inputValue === 'string';

const newOptionComputationDependencies: $Keys<CreatableProps>[] = ['isValidNewOption', 'inputValue', 'value', 'options'];
const optionsComputationDependencies: $Keys<CreatableProps>[] = ['options', 'createOptionPosition'];
const createMemoizedOptionsBuilder = () => {
  let prevProps: CreatableProps = {};
  let cachedNewOption: ?OptionType;
  let cachedOptions: ?OptionsType;

  const getMemoizedOptions = (props: CreatableProps): OptionsType => {
    const {
      allowCreateWhileLoading,
      createOptionPosition,
      formatCreateLabel,
      getNewOptionData,
      inputValue,
      isLoading,
      isValidNewOption,
      value,
      options = [],
    } = props;

    let newOption = cachedNewOption;
    if (!newOption || newOptionComputationDependencies.some(key => prevProps[key] !== props[key])) {
      newOption = isValidNewOption(inputValue, cleanValue(value), options)
        ? getNewOptionData(inputValue, formatCreateLabel(inputValue))
        : undefined;
    }

    if (
      !cachedOptions ||
      newOption !== cachedNewOption ||
      (!allowCreateWhileLoading && prevProps.isLoading !== isLoading) || // only check isLoading in case allowCreateWhileLoading is false
      (allowCreateWhileLoading !== prevProps.allowCreateWhileLoading && (allowCreateWhileLoading ? prevProps.isLoading : isLoading)) || // a change in createWhileLoading only requires a new computation in case the result of allowCreateWhileLoading || !isLoading changed
      optionsComputationDependencies.some(key => prevProps[key] !== props[key])
    ) {
      cachedOptions = (allowCreateWhileLoading || !isLoading) && newOption
        ? createOptionPosition === 'first'
          ? [newOption, ...options]
          : [...options, newOption]
        : options;
    }

    cachedNewOption = newOption;
    return cachedOptions;
  };
  getMemoizedOptions.getNewOption = () => cachedNewOption;

  return getMemoizedOptions;
};