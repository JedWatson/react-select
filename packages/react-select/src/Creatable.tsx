import React, { MutableRefObject, ReactElement, RefAttributes } from 'react';
import Select from './Select';
import { GroupBase } from './types';
import useStateManager, { StateManagerProps } from './useStateManager';
import useCreatable, { CreatableAdditionalProps } from './useCreatable';

export type CreatableProps<
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
> = StateManagerProps<Option, IsMulti, Group> &
  CreatableAdditionalProps<Option, Group>;

type CreatableSelect = <
  Option = unknown,
  IsMulti extends boolean = false,
  Group extends GroupBase<Option> = GroupBase<Option>
>(
  props: CreatableProps<Option, IsMulti, Group> &
    RefAttributes<Select<Option, IsMulti, Group>>
) => ReactElement;

const CreatableSelect = React.forwardRef(
  <Option, IsMulti extends boolean, Group extends GroupBase<Option>>(
    props: CreatableProps<Option, IsMulti, Group>,
    ref:
      | ((instance: Select<Option, IsMulti, Group> | null) => void)
      | MutableRefObject<Select<Option, IsMulti, Group> | null>
      | null
  ) => {
    const creatableProps = useStateManager(props);
    const selectProps = useCreatable(creatableProps);

    return <Select ref={ref} {...selectProps} />;
  }
) as CreatableSelect;

export default CreatableSelect;
