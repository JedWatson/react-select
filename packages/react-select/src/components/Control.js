// @flow
/** @jsx jsx */
import { type Node, type ElementRef } from 'react';
import { jsx } from '@emotion/core';

import type { CommonProps, PropsWithStyles } from '../types';

type State = {
  /** Whether the select is disabled. */
  isDisabled: boolean,
  /** Whether the select is focused. */
  isFocused: boolean,
  /** Whether the select is expanded. */
  menuIsOpen: boolean,
};

export type ControlProps = CommonProps &
  PropsWithStyles &
  State & {
    /** Children to render. */
    children: Node,
    innerRef: ElementRef<*>,
    /** The mouse down event and the innerRef to pass down to the controller element. */
    innerProps: {
      onMouseDown: (SyntheticMouseEvent<HTMLElement>) => void,
    },
  };

export const css = ({
  isDisabled,
  isFocused,
  theme: { colors, borderRadius, spacing },
}: ControlProps) => ({
  label: 'control',
  alignItems: 'center',
  backgroundColor: isDisabled ? colors.neutral5 : colors.neutral0,
  borderColor: isDisabled
    ? colors.neutral10
    : isFocused
    ? colors.primary
    : colors.neutral20,
  borderRadius: borderRadius,
  borderStyle: 'solid',
  borderWidth: 1,
  boxShadow: isFocused ? `0 0 0 1px ${colors.primary}` : null,
  cursor: 'default',
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'space-between',
  minHeight: spacing.controlHeight,
  outline: '0 !important',
  position: 'relative',
  transition: 'all 100ms',

  '&:hover': {
    borderColor: isFocused ? colors.primary : colors.neutral30,
  },
});

const Control = (props: ControlProps) => {
  const {
    children,
    cx,
    getStyles,
    className,
    isDisabled,
    isFocused,
    innerRef,
    innerProps,
    menuIsOpen,
  } = props;
  return (
    <div
      ref={innerRef}
      css={getStyles('control', props)}
      className={cx(
        {
          control: true,
          'control--is-disabled': isDisabled,
          'control--is-focused': isFocused,
          'control--menu-is-open': menuIsOpen,
        },
        className
      )}
      {...innerProps}
    >
      {children}
    </div>
  );
};

export default Control;
