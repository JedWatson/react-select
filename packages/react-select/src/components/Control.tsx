/** @jsx jsx */
import { ReactNode, Ref } from 'react';
import { jsx } from '@emotion/react';

import {
  CommonPropsAndClassName,
  CSSObjectWithLabel,
  GroupBase,
} from '../types';

export interface ControlProps<
  Option = unknown,
  IsMulti extends boolean = boolean,
  Group extends GroupBase<Option> = GroupBase<Option>
> extends CommonPropsAndClassName<Option, IsMulti, Group> {
  /** Children to render. */
  children: ReactNode;
  innerRef: Ref<HTMLDivElement>;
  /** The mouse down event and the innerRef to pass down to the controller element. */
  innerProps: JSX.IntrinsicElements['div'];
  /** Whether the select is disabled. */
  isDisabled: boolean;
  /** Whether the select is focused. */
  isFocused: boolean;
  /** Whether the select is expanded. */
  menuIsOpen: boolean;
}

export const css = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>({
  isDisabled,
  isFocused,
  theme: { colors, borderRadius, spacing },
}: ControlProps<Option, IsMulti, Group>): CSSObjectWithLabel => ({
  label: 'control',
  alignItems: 'center',
  backgroundColor: isDisabled ? colors.neutral5 : colors.neutral0,
  borderColor: isDisabled
    ? colors.neutral10
    : isFocused
    ? colors.primary
    : colors.neutral20,
  borderRadius: borderRadius,
  borderStyle: 'solid',
  borderWidth: 1,
  boxShadow: isFocused ? `0 0 0 1px ${colors.primary}` : undefined,
  cursor: 'default',
  display: 'flex',
  flexWrap: 'wrap',
  justifyContent: 'space-between',
  minHeight: spacing.controlHeight,
  outline: '0 !important',
  position: 'relative',
  transition: 'all 100ms',

  '&:hover': {
    borderColor: isFocused ? colors.primary : colors.neutral30,
  },
});

const Control = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>(
  props: ControlProps<Option, IsMulti, Group>
) => {
  const {
    children,
    cx,
    getStyles,
    className,
    isDisabled,
    isFocused,
    innerRef,
    innerProps,
    menuIsOpen,
  } = props;
  return (
    <div
      ref={innerRef}
      css={getStyles('control', props)}
      className={cx(
        {
          control: true,
          'control--is-disabled': isDisabled,
          'control--is-focused': isFocused,
          'control--menu-is-open': menuIsOpen,
        },
        className
      )}
      {...innerProps}
    >
      {children}
    </div>
  );
};

export default Control;
