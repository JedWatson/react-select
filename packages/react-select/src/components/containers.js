// @flow
/** @jsx jsx */
import { type Node } from 'react';
import { jsx } from '@emotion/core';
import type { CommonProps, KeyboardEventHandler } from '../types';

// ==============================
// Root Container
// ==============================

type ContainerState = {
  /** Whether the select is disabled. */
  isDisabled: boolean,
  /** Whether the text in the select is indented from right to left. */
  isRtl: boolean,
};

export type ContainerProps = CommonProps &
  ContainerState & {
    /** The children to be rendered. */
    children: Node,
    /** Inner props to be passed down to the container. */
    innerProps: { onKeyDown: KeyboardEventHandler },
  };
export const containerCSS = ({ isDisabled, isRtl }: ContainerState) => ({
  label: 'container',
  direction: isRtl ? 'rtl' : null,
  pointerEvents: isDisabled ? 'none' : null, // cancel mouse events when disabled
  position: 'relative',
});
export const SelectContainer = (props: ContainerProps) => {
  const {
    children,
    className,
    cx,
    getStyles,
    innerProps,
    isDisabled,
    isRtl,
  } = props;
  return (
    <div
      css={getStyles('container', props)}
      className={cx(
        {
          '--is-disabled': isDisabled,
          '--is-rtl': isRtl,
        },
        className
      )}
      {...innerProps}
    >
      {children}
    </div>
  );
};

// ==============================
// Value Container
// ==============================

export type ValueContainerProps = CommonProps & {
  /** Set when the value container should hold multiple values */
  isMulti: boolean,
  /** Whether the value container currently holds a value. */
  hasValue: boolean,
  /** The children to be rendered. */
  children: Node,
};
export const valueContainerCSS = ({
  theme: { spacing },
}: ValueContainerProps) => ({
  alignItems: 'center',
  display: 'flex',
  flex: 1,
  flexWrap: 'wrap',
  padding: `${spacing.baseUnit / 2}px ${spacing.baseUnit * 2}px`,
  WebkitOverflowScrolling: 'touch',
  position: 'relative',
  overflow: 'hidden',
});
export const ValueContainer = (props: ValueContainerProps) => {
  const { children, className, cx, isMulti, getStyles, hasValue } = props;

  return (
    <div
      css={getStyles('valueContainer', props)}
      className={cx(
        {
          'value-container': true,
          'value-container--is-multi': isMulti,
          'value-container--has-value': hasValue,
        },
        className
      )}
    >
      {children}
    </div>
  );
};

// ==============================
// Indicator Container
// ==============================

type IndicatorsState = {
  /** Whether the text should be rendered right to left. */
  isRtl: boolean,
};

export type IndicatorContainerProps = CommonProps &
  IndicatorsState & {
    /** The children to be rendered. */
    children: Node,
  };

export const indicatorsContainerCSS = () => ({
  alignItems: 'center',
  alignSelf: 'stretch',
  display: 'flex',
  flexShrink: 0,
});
export const IndicatorsContainer = (props: IndicatorContainerProps) => {
  const { children, className, cx, getStyles } = props;

  return (
    <div
      css={getStyles('indicatorsContainer', props)}
      className={cx(
        {
          indicators: true,
        },
        className
      )}
    >
      {children}
    </div>
  );
};
