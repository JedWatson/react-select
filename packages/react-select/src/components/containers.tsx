/** @jsx jsx */
import { ReactNode } from 'react';
import { jsx } from '@emotion/react';
import {
  CommonPropsAndClassName,
  CSSObjectWithLabel,
  GroupBase,
} from '../types';

// ==============================
// Root Container
// ==============================

export interface ContainerProps<
  Option = unknown,
  IsMulti extends boolean = boolean,
  Group extends GroupBase<Option> = GroupBase<Option>
> extends CommonPropsAndClassName<Option, IsMulti, Group> {
  /** Whether the select is disabled. */
  isDisabled: boolean;
  isFocused: boolean;
  /** The children to be rendered. */
  children: ReactNode;
  /** Inner props to be passed down to the container. */
  innerProps: JSX.IntrinsicElements['div'];
}
export const containerCSS = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>({
  isDisabled,
  isRtl,
}: ContainerProps<Option, IsMulti, Group>): CSSObjectWithLabel => ({
  label: 'container',
  direction: isRtl ? 'rtl' : undefined,
  pointerEvents: isDisabled ? 'none' : undefined, // cancel mouse events when disabled
  position: 'relative',
});
export const SelectContainer = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>(
  props: ContainerProps<Option, IsMulti, Group>
) => {
  const { children, className, cx, getStyles, innerProps, isDisabled, isRtl } =
    props;
  return (
    <div
      css={getStyles('container', props)}
      className={cx(
        {
          '--is-disabled': isDisabled,
          '--is-rtl': isRtl,
        },
        className
      )}
      {...innerProps}
    >
      {children}
    </div>
  );
};

// ==============================
// Value Container
// ==============================

export interface ValueContainerProps<
  Option = unknown,
  IsMulti extends boolean = boolean,
  Group extends GroupBase<Option> = GroupBase<Option>
> extends CommonPropsAndClassName<Option, IsMulti, Group> {
  /** Props to be passed to the value container element. */
  innerProps?: JSX.IntrinsicElements['div'];
  /** The children to be rendered. */
  children: ReactNode;
  isDisabled: boolean;
}
export const valueContainerCSS = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>({
  theme: { spacing },
  isMulti,
  hasValue,
}: ValueContainerProps<Option, IsMulti, Group>): CSSObjectWithLabel => ({
  alignItems: 'center',
  display: isMulti && hasValue ? 'flex' : 'grid',
  flex: 1,
  flexWrap: 'wrap',
  padding: `${spacing.baseUnit / 2}px ${spacing.baseUnit * 2}px`,
  WebkitOverflowScrolling: 'touch',
  position: 'relative',
  overflow: 'hidden',
});
export const ValueContainer = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>(
  props: ValueContainerProps<Option, IsMulti, Group>
) => {
  const { children, className, cx, innerProps, isMulti, getStyles, hasValue } =
    props;

  return (
    <div
      css={getStyles('valueContainer', props)}
      className={cx(
        {
          'value-container': true,
          'value-container--is-multi': isMulti,
          'value-container--has-value': hasValue,
        },
        className
      )}
      {...innerProps}
    >
      {children}
    </div>
  );
};

// ==============================
// Indicator Container
// ==============================

export interface IndicatorsContainerProps<
  Option = unknown,
  IsMulti extends boolean = boolean,
  Group extends GroupBase<Option> = GroupBase<Option>
> extends CommonPropsAndClassName<Option, IsMulti, Group> {
  isDisabled: boolean;
  /** The children to be rendered. */
  children: ReactNode;
  /** Props to be passed to the indicators container element. */
  innerProps?: {};
}

export const indicatorsContainerCSS = (): CSSObjectWithLabel => ({
  alignItems: 'center',
  alignSelf: 'stretch',
  display: 'flex',
  flexShrink: 0,
});
export const IndicatorsContainer = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>(
  props: IndicatorsContainerProps<Option, IsMulti, Group>
) => {
  const { children, className, cx, innerProps, getStyles } = props;

  return (
    <div
      css={getStyles('indicatorsContainer', props)}
      className={cx(
        {
          indicators: true,
        },
        className
      )}
      {...innerProps}
    >
      {children}
    </div>
  );
};
