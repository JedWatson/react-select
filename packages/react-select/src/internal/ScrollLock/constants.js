export const STYLE_KEYS = [
  'boxSizing',
  'height',
  'overflow',
  'paddingRight',
  'position'
];

export const LOCK_STYLES = {
  boxSizing: 'border-box', // account for possible declaration `width: 100%;` on body
  overflow: 'hidden',
  position: 'relative',
  height: '100%'
};
