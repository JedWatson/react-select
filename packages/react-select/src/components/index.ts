import { ComponentType } from 'react';
import {
  ContainerProps,
  IndicatorsContainer,
  IndicatorsContainerProps,
  SelectContainer,
  ValueContainer,
  ValueContainerProps,
} from './containers';
import {
  ClearIndicator,
  ClearIndicatorProps,
  CrossIcon,
  CrossIconProps,
  DownChevron,
  DownChevronProps,
  DropdownIndicator,
  DropdownIndicatorProps,
  IndicatorSeparator,
  IndicatorSeparatorProps,
  LoadingIndicator,
  LoadingIndicatorProps,
} from './indicators';

import Control, { ControlProps } from './Control';
import Group, { GroupHeading, GroupHeadingProps, GroupProps } from './Group';
import Input, { InputProps } from './Input';
import Menu, {
  LoadingMessage,
  MenuList,
  MenuListProps,
  MenuPortal,
  MenuPortalProps,
  MenuProps,
  NoOptionsMessage,
  NoticeProps,
} from './Menu';
import MultiValue, {
  MultiValueContainer,
  MultiValueGenericProps,
  MultiValueLabel,
  MultiValueProps,
  MultiValueRemove,
  MultiValueRemoveProps,
} from './MultiValue';
import Option, { OptionProps } from './Option';
import Placeholder, { PlaceholderProps } from './Placeholder';
import SingleValue, { SingleValueProps } from './SingleValue';
import { GroupBase } from '../types';

export interface SelectComponents<
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
> {
  ClearIndicator: ComponentType<ClearIndicatorProps<Option, IsMulti, Group>>;
  Control: ComponentType<ControlProps<Option, IsMulti, Group>>;
  DropdownIndicator: ComponentType<
    DropdownIndicatorProps<Option, IsMulti, Group>
  > | null;
  DownChevron: ComponentType<DownChevronProps>;
  CrossIcon: ComponentType<CrossIconProps>;
  Group: ComponentType<GroupProps<Option, IsMulti, Group>>;
  GroupHeading: ComponentType<GroupHeadingProps<Option, IsMulti, Group>>;
  IndicatorsContainer: ComponentType<
    IndicatorsContainerProps<Option, IsMulti, Group>
  >;
  IndicatorSeparator: ComponentType<
    IndicatorSeparatorProps<Option, IsMulti, Group>
  > | null;
  Input: ComponentType<InputProps<Option, IsMulti, Group>>;
  LoadingIndicator: ComponentType<
    LoadingIndicatorProps<Option, IsMulti, Group>
  >;
  Menu: ComponentType<MenuProps<Option, IsMulti, Group>>;
  MenuList: ComponentType<MenuListProps<Option, IsMulti, Group>>;
  MenuPortal: ComponentType<MenuPortalProps<Option, IsMulti, Group>>;
  LoadingMessage: ComponentType<NoticeProps<Option, IsMulti, Group>>;
  NoOptionsMessage: ComponentType<NoticeProps<Option, IsMulti, Group>>;
  MultiValue: ComponentType<MultiValueProps<Option, IsMulti, Group>>;
  MultiValueContainer: ComponentType<
    MultiValueGenericProps<Option, IsMulti, Group>
  >;
  MultiValueLabel: ComponentType<
    MultiValueGenericProps<Option, IsMulti, Group>
  >;
  MultiValueRemove: ComponentType<
    MultiValueRemoveProps<Option, IsMulti, Group>
  >;
  Option: ComponentType<OptionProps<Option, IsMulti, Group>>;
  Placeholder: ComponentType<PlaceholderProps<Option, IsMulti, Group>>;
  SelectContainer: ComponentType<ContainerProps<Option, IsMulti, Group>>;
  SingleValue: ComponentType<SingleValueProps<Option, IsMulti, Group>>;
  ValueContainer: ComponentType<ValueContainerProps<Option, IsMulti, Group>>;
}

export type SelectComponentsConfig<
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
> = Partial<SelectComponents<Option, IsMulti, Group>>;

export const components = {
  ClearIndicator: ClearIndicator,
  Control: Control,
  DropdownIndicator: DropdownIndicator,
  DownChevron: DownChevron,
  CrossIcon: CrossIcon,
  Group: Group,
  GroupHeading: GroupHeading,
  IndicatorsContainer: IndicatorsContainer,
  IndicatorSeparator: IndicatorSeparator,
  Input: Input,
  LoadingIndicator: LoadingIndicator,
  Menu: Menu,
  MenuList: MenuList,
  MenuPortal: MenuPortal,
  LoadingMessage: LoadingMessage,
  NoOptionsMessage: NoOptionsMessage,
  MultiValue: MultiValue,
  MultiValueContainer: MultiValueContainer,
  MultiValueLabel: MultiValueLabel,
  MultiValueRemove: MultiValueRemove,
  Option: Option,
  Placeholder: Placeholder,
  SelectContainer: SelectContainer,
  SingleValue: SingleValue,
  ValueContainer: ValueContainer,
};

export type SelectComponentsGeneric = typeof components;

interface Props<
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
> {
  components: SelectComponentsConfig<Option, IsMulti, Group>;
}

export const defaultComponents = <
  Option,
  IsMulti extends boolean,
  Group extends GroupBase<Option>
>(
  props: Props<Option, IsMulti, Group>
): SelectComponentsGeneric =>
  ({
    ...components,
    ...props.components,
  } as SelectComponentsGeneric);
