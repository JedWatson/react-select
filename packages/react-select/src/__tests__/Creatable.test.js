import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import cases from 'jest-in-case';

import Creatable from '../Creatable';
import { OPTIONS } from './constants';
import { components } from '../components';
const { Menu, NoOptionsMessage } = components;

test('defaults - snapshot', () => {
  const tree = shallow(<Creatable />);
  expect(toJson(tree)).toMatchSnapshot();
});

cases('filtered option is an exact match for an existing option',
  ({ props = { options: OPTIONS } }) => {
    const creatableSelectWrapper = mount(<Creatable menuIsOpen {...props} />);
    creatableSelectWrapper.setProps({ inputValue: 'one' });
    expect(creatableSelectWrapper.find(Menu).text()).not.toEqual(
      expect.stringContaining('create')
    );
  },
  {
    'single select > should not show "create..." prompt"': {},
    'multi select > should not show "create..." prompt"': {
      props: {
        isMulti: true,
        options: OPTIONS,
      },
    },
  }
);

cases('filterOptions returns invalid value ( null )',
  ({ props = { option: OPTIONS } }) => {
    let filterOptionSpy = jest.fn().mockReturnValue(null);
    const creatableSelectWrapper = mount(
      <Creatable filterOption={filterOptionSpy} menuIsOpen {...props} />
    );
    creatableSelectWrapper.setProps({ inputValue: 'one' });
    expect(creatableSelectWrapper.find(NoOptionsMessage).exists()).toBeTruthy();
    expect(creatableSelectWrapper.find(Menu).text()).not.toEqual(
      expect.stringContaining('create')
    );
  },
  {
    'single select > should not show "create..." prompt"': {},
    'multi select > should not show "create..." prompt"': {
      props: {
        isMulti: true,
        option: OPTIONS,
      },
    },
  }
);

cases('inputValue does not match any option after filter',
  ({ props = { options: OPTIONS } }) => {
    const creatableSelectWrapper = mount(<Creatable menuIsOpen {...props} />);
    creatableSelectWrapper.setProps({ inputValue: 'option not is list' });
    expect(creatableSelectWrapper.find(Menu).text()).toBe(
      'Create "option not is list"'
    );
  },
  {
    'single select > should show a placeholder "create..." prompt': {},
    'multi select > should show a placeholder "create..." prompt': {
      props: {
        isMulti: true,
        options: OPTIONS,
      },
    },
  }
);

cases('isValidNewOption() prop',
  ({ props = { options: OPTIONS } }) => {
    let isValidNewOption = jest.fn(options => options === 'new Option');
    const creatableSelectWrapper = mount(
      <Creatable menuIsOpen {...props} isValidNewOption={isValidNewOption} />
    );

    creatableSelectWrapper.setProps({ inputValue: 'new Option' });
    expect(creatableSelectWrapper.find(Menu).text()).toEqual(
      expect.stringContaining('Create "new Option"')
    );
    expect(creatableSelectWrapper.find(NoOptionsMessage).exists()).toBeFalsy();

    creatableSelectWrapper.setProps({ inputValue: 'invalid new Option' });
    expect(creatableSelectWrapper.find(Menu).text()).not.toEqual(
      expect.stringContaining('Create "invalid new Option"')
    );
    expect(creatableSelectWrapper.find(NoOptionsMessage).exists()).toBeTruthy();
  },
  {
    'single select > should show "create..." prompt only if isValidNewOption returns thruthy value': {},
    'multi select > should show "create..." prompt only if isValidNewOption returns thruthy value': {
      props: {
        isMulti: true,
        options: OPTIONS,
      },
    },
  }
);

cases('close by hitting escape with search text present',
  ({ props = { options: OPTIONS } }) => {
    const creatableSelectWrapper = mount(<Creatable menuIsOpen {...props} />);
    creatableSelectWrapper.setState({ inputValue: 'new Option' });
    creatableSelectWrapper.update();
    creatableSelectWrapper.simulate('keyDown', { keyCode: 27, key: 'Escape' });
    creatableSelectWrapper.update();
    expect(creatableSelectWrapper.state('inputValue')).toBe('');
  },
  {
    'single select > should remove the search text': {
    },
    'multi select > should remove the search text': {
      props: {
        isMulti: true,
        options: OPTIONS,
      },
    },
  }
);

test('should remove the new option after closing on blur', () => {
  const creatableSelectWrapper = mount(
    <Creatable menuIsOpen options={OPTIONS} />
  );
  creatableSelectWrapper.setState({ inputValue: 'new Option' });
  creatableSelectWrapper.find('Control input').simulate('blur');
  expect(creatableSelectWrapper.state('inputValue')).toBe('');
});

cases('getNewOptionData() prop',
  ({ props = { options: OPTIONS } }) => {
    let getNewOptionDataSpy = jest.fn(label => ({
      label: `custom text ${label}`,
      value: label,
    }));
    const creatableSelectWrapper = mount(
      <Creatable menuIsOpen {...props} getNewOptionData={getNewOptionDataSpy} />
    );
    creatableSelectWrapper.setState({ inputValue: 'new Option' });
    expect(creatableSelectWrapper.find(Menu).text()).toEqual(
      expect.stringContaining('custom text new Option')
    );
  },
  {
    'single select > should create option as per label returned from getNewOptionData': {
    },
    'multi select > should create option as per label returned from getNewOptionData': {
      props: {
        isMulti: true,
        options: OPTIONS,
      },
    },
  }
);

cases('formatCreateLabel() prop',
  ({ props = { options: OPTIONS } }) => {
    let formatCreateLabelSpy = jest.fn(label => `custom label "${label}"`);
    const creatableSelectWrapper = mount(
      <Creatable
        menuIsOpen
        {...props}
        formatCreateLabel={formatCreateLabelSpy}
      />
    );
    creatableSelectWrapper.setState({ inputValue: 'new Option' });
    expect(creatableSelectWrapper.find(Menu).text()).toEqual(
      expect.stringContaining('custom label "new Option"')
    );
  },
  {
    'single select > should show label of custom option as per text returned from formatCreateLabel': {},
    'multi select > should show label of custom option as per text returned from formatCreateLabel': {
      props: {
        isMulti: true,
        options: OPTIONS,
      },
    },
  }
);
