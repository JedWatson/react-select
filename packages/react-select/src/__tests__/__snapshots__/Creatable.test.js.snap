// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults - snapshot 1`] = `
<Creatable
  allowCreateWhileLoading={false}
  createOptionPosition="last"
  formatCreateLabel={[Function]}
  getNewOptionData={[Function]}
  inputValue=""
  isValidNewOption={[Function]}
  menuIsOpen={false}
  onChange={[Function]}
  onInputChange={[Function]}
  onMenuClose={[Function]}
  onMenuOpen={[Function]}
  value={null}
/>
`;
